File: concept/answer/ConceptMapImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept.answer;

import com.vaticle.typedb.client.api.answer.ConceptMap;
import com.vaticle.typedb.client.api.concept.Concept;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.client.concept.ConceptImpl;
import com.vaticle.typedb.common.collection.Pair;
import com.vaticle.typedb.protocol.AnswerProto;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.NONEXISTENT_EXPLAINABLE_CONCEPT;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.NONEXISTENT_EXPLAINABLE_OWNERSHIP;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Query.VARIABLE_DOES_NOT_EXIST;

public class ConceptMapImpl implements ConceptMap {

    private final Map<String, Concept> map;
    private final Explainables explainables;

    public ConceptMapImpl(Map<String, Concept> map) {
        this(map, new ExplainablesImpl());
    }

    public ConceptMapImpl(Map<String, Concept> map, Explainables explainables) {
        this.map = Collections.unmodifiableMap(map);
        this.explainables = explainables;
    }

    public static ConceptMap of(AnswerProto.ConceptMap res) {
        Map<String, Concept> variableMap = new HashMap<>();
        res.getMapMap().forEach((resVar, resConcept) -> variableMap.put(resVar, ConceptImpl.of(resConcept)));
        return new ConceptMapImpl(Collections.unmodifiableMap(variableMap), of(res.getExplainables()));
    }

    private static Explainables of(AnswerProto.Explainables explainables) {
        Map<String, Explainable> relations = new HashMap<>();
        explainables.getRelationsMap().forEach((var, explainable) -> {
            relations.put(var, ExplainableImpl.of(explainable));
        });
        Map<String, Explainable> attributes = new HashMap<>();
        explainables.getAttributesMap().forEach((var, explainable) -> {
            attributes.put(var, ExplainableImpl.of(explainable));
        });
        Map<Pair<String, String>, Explainable> ownerships = new HashMap<>();
        explainables.getOwnershipsMap().forEach((var, ownedMap) -> {
            ownedMap.getOwnedMap().forEach((owned, explainable) -> {
                ownerships.put(new Pair<>(var, owned), ExplainableImpl.of(explainable));
            });
        });
        return new ExplainablesImpl(relations, attributes, ownerships);
    }

    @Override
    public Map<String, Concept> map() {
        return map;
    }

    @Override
    public Collection<Concept> concepts() {
        return map.values();
    }

    @Override
    public Concept get(String variable) {
        Concept concept = map.get(variable);
        if (concept == null) throw new TypeDBClientException(VARIABLE_DOES_NOT_EXIST, variable);
        return concept;
    }

    @Override
    public Explainables explainables() {
        return explainables;
    }

    @Override
    public String toString() {
        return map.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .map(e -> "[" + e.getKey() + "/" + e.getValue() + "]").collect(Collectors.joining());
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ConceptMapImpl a2 = (ConceptMapImpl) obj;
        return map.equals(a2.map);
    }

    @Override
    public int hashCode() {
        return map.hashCode();
    }

    public static class ExplainablesImpl implements Explainables {

        Map<String, Explainable> explainableRelations;
        Map<String, Explainable> explainableAttributes;
        Map<Pair<String, String>, Explainable> explainableOwnerships;

        ExplainablesImpl() {
            this(new HashMap<>(), new HashMap<>(), new HashMap<>());
        }

        ExplainablesImpl(Map<String, Explainable> explainableRelations, Map<String, Explainable> explainableAttributes,
                         Map<Pair<String, String>, Explainable> explainableOwnerships) {
            this.explainableRelations = explainableRelations;
            this.explainableAttributes = explainableAttributes;
            this.explainableOwnerships = explainableOwnerships;
        }

        @Override
        public Explainable relation(String variable) {
            Explainable explainable = explainableRelations.get(variable);
            if (explainable == null) throw new TypeDBClientException(NONEXISTENT_EXPLAINABLE_CONCEPT, variable);
            return explainable;
        }

        @Override
        public Explainable attribute(String variable) {
            Explainable explainable = explainableAttributes.get(variable);
            if (explainable == null) throw new TypeDBClientException(NONEXISTENT_EXPLAINABLE_CONCEPT, variable);
            return explainable;
        }

        @Override
        public Explainable ownership(String owner, String attribute) {
            Explainable explainable = explainableOwnerships.get(new Pair<>(owner, attribute));
            if (explainable == null)
                throw new TypeDBClientException(NONEXISTENT_EXPLAINABLE_OWNERSHIP, owner, attribute);
            return explainable;
        }

        @Override
        public Map<String, Explainable> relations() {
            return this.explainableRelations;
        }

        @Override
        public Map<String, Explainable> attributes() {
            return this.explainableAttributes;
        }

        @Override
        public Map<Pair<String, String>, Explainable> ownerships() {
            return this.explainableOwnerships;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ExplainablesImpl that = (ExplainablesImpl) o;
            return explainableRelations.equals(that.explainableRelations) &&
                    explainableAttributes.equals(that.explainableAttributes) &&
                    explainableOwnerships.equals(that.explainableOwnerships);
        }

        @Override
        public int hashCode() {
            return Objects.hash(explainableRelations, explainableAttributes, explainableOwnerships);
        }

    }

    static class ExplainableImpl implements Explainable {

        private final String conjunction;
        private final long id;

        ExplainableImpl(String conjunction, long id) {
            this.conjunction = conjunction;
            this.id = id;
        }

        public static Explainable of(AnswerProto.Explainable explainable) {
            return new ExplainableImpl(explainable.getConjunction(), explainable.getId());
        }

        @Override
        public String conjunction() {
            return conjunction;
        }

        @Override
        public long id() {
            return id;
        }

        @Override
        public boolean equals(final Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            final ExplainableImpl that = (ExplainableImpl) o;
            return id == that.id;
        }

        @Override
        public int hashCode() {
            return (int) id;
        }
    }
}

File: concept/answer/ConceptMapImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { ConceptMap as ConceptMapProto, Explainable as ExplainableProto, Explainables, Explainables as ExplainablesProto } from "typedb-protocol/common/answer_pb";
import { Concept as ConceptProto } from "typedb-protocol/common/concept_pb";
import { ConceptMap } from "../../api/answer/ConceptMap";
import { Concept } from "../../api/concept/Concept";
import { ErrorMessage } from "../../common/errors/ErrorMessage";
import { TypeDBClientError } from "../../common/errors/TypeDBClientError";
import { ThingImpl, TypeImpl } from "../../dependencies_internal";
import {ValueImpl} from "../value/ValueImpl";

export class ConceptMapImpl implements ConceptMap {

    private readonly _concepts: Map<string, Concept>;
    private readonly _explainables: ConceptMap.Explainables;

    constructor(concepts: Map<string, Concept>, explainables: ConceptMap.Explainables) {
        this._concepts = concepts;
        this._explainables = explainables;
    }

    concepts(): IterableIterator<Concept> {
        return this._concepts.values();
    }

    get(variable: string): Concept {
        return this._concepts.get(variable);
    }

    get map(): Map<string, Concept> {
        return this._concepts;
    }

    get explainables(): ConceptMap.Explainables {
        return this._explainables;
    }

    toJSONRecord(): Record<string, Record<string, boolean | string | number>> {
        const object: Record<string, Record<string, boolean | string | number>> = {}
        for (const key of this._concepts.keys()) {
            object[key] = this._concepts.get(key).toJSONRecord();
        }
        return object;
    }
}

/* eslint no-inner-declarations: "off" */
export namespace ConceptMapImpl {

    import NONEXISTENT_EXPLAINABLE_CONCEPT = ErrorMessage.Query.NONEXISTENT_EXPLAINABLE_CONCEPT;
    import NONEXISTENT_EXPLAINABLE_OWNERSHIP = ErrorMessage.Query.NONEXISTENT_EXPLAINABLE_OWNERSHIP;
    import Owned = Explainables.Owned;

    export function of(proto: ConceptMapProto): ConceptMap {
        const variableMap = new Map<string, Concept>();
        proto.getMapMap().forEach((protoConcept: ConceptProto, resLabel: string) => {
            let concept;
            if (protoConcept.hasThing()) concept = ThingImpl.of(protoConcept.getThing());
            else if (protoConcept.hasType()) concept = TypeImpl.of(protoConcept.getType());
            else concept = ValueImpl.of(protoConcept.getValue());
            variableMap.set(resLabel, concept);
        });
        const explainables = proto.hasExplainables() ? ofExplainables(proto.getExplainables()) : emptyExplainables();
        return new ConceptMapImpl(variableMap, explainables);
    }

    function ofExplainables(proto: ExplainablesProto): ConceptMap.Explainables {
        const relations = new Map<string, ConceptMap.Explainable>();
        proto.getRelationsMap().forEach((explainable: ExplainableProto, variable: string) =>
            relations.set(variable, ofExplainable(explainable))
        );
        const attributes = new Map<string, ConceptMap.Explainable>();
        proto.getAttributesMap().forEach((explainable: ExplainableProto, variable: string) =>
            relations.set(variable, ofExplainable(explainable))
        );
        const ownerships = new Map<[string, string], ConceptMap.Explainable>();
        proto.getOwnershipsMap().forEach((owned: Owned, owner: string) =>
            owned.getOwnedMap().forEach((explainable: ExplainableProto, attribute: string) => {
                ownerships.set([owner, attribute], ofExplainable(explainable))
            })
        );
        return new ExplainablesImpl(relations, attributes, ownerships)
    }

    function emptyExplainables() {
        return new ExplainablesImpl(
            new Map<string, ConceptMap.Explainable>(),
            new Map<string, ConceptMap.Explainable>(),
            new Map<[string, string], ConceptMap.Explainable>()
        );
    }

    function ofExplainable(proto: ExplainableProto): ConceptMap.Explainable {
        return new ExplainableImpl(proto.getConjunction(), proto.getId());
    }

    export class ExplainablesImpl implements ConceptMap.Explainables {
        private readonly _relations: Map<string, ConceptMap.Explainable>;
        private readonly _attributes: Map<string, ConceptMap.Explainable>;
        private readonly _ownerships: Map<[string, string], ConceptMap.Explainable>;

        constructor(relations: Map<string, ConceptMap.Explainable>, attributes: Map<string, ConceptMap.Explainable>,
                    ownerships: Map<[string, string], ConceptMap.Explainable>) {
            this._relations = relations;
            this._attributes = attributes;
            this._ownerships = ownerships;
        }

        relation(variable: string): ConceptMap.Explainable {
            const explainable = this._relations.get(variable);
            if (!explainable) throw new TypeDBClientError(NONEXISTENT_EXPLAINABLE_CONCEPT.message(variable));
            return explainable;
        }

        attribute(variable: string): ConceptMap.Explainable {
            const explainable = this._attributes.get(variable);
            if (!explainable) throw new TypeDBClientError(NONEXISTENT_EXPLAINABLE_CONCEPT.message(variable));
            return explainable;
        }

        ownership(owner: string, attribute: string): ConceptMap.Explainable {
            for (const entry of this._ownerships) {
                if (entry[0][0] === owner && entry[0][1] === attribute) return entry[1];
            }
            throw new TypeDBClientError(NONEXISTENT_EXPLAINABLE_OWNERSHIP.message(owner, attribute));
        }

        get relations(): Map<string, ConceptMap.Explainable> {
            return this._relations;
        }

        get attributes(): Map<string, ConceptMap.Explainable> {
            return this._attributes;
        }

        get ownerships(): Map<[string, string], ConceptMap.Explainable> {
            return this._ownerships;
        }
    }

    export class ExplainableImpl implements ConceptMap.Explainable {
        private readonly _conjunction: string;
        private readonly _id: number;

        constructor(conjunction: string, id: number) {
            this._conjunction = conjunction;
            this._id = id;
        }

        get conjunction(): string {
            return this._conjunction;
        }

        get id(): number {
            return this._id;
        }
    }
}
